File: bedrock_llm/bedrock_llm_extension.go
/**
 *
 * Agora Real Time Engagement
 * Created by lixinhui in 2024.
 * Copyright (c) 2024 Agora IO. All rights reserved.
 *
 */
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package extension

import (
	"fmt"
	"log/slog"
	"sync"
	"sync/atomic"
	"time"

	"agora.io/rte/rtego"
	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime/types"
)

var (
	logTag = slog.String("extension", "BEDROCK_LLM_EXTENSION")
)

type bedrockLLMExtension struct {
	rtego.DefaultExtension
	bedrockLLM *bedrockLLM
}

const (
	cmdInFlush                              = "flush"
	cmdOutFlush                             = "flush"
	dataInTextDataPropertyText              = "text"
	dataInTextDataPropertyIsFinal           = "is_final"
	dataOutTextDataPropertyText             = "text"
	dataOutTextDataPropertyTextEndOfSegment = "end_of_segment"

	propertyRegion          = "region"            // Optional
	propertyAccessKey       = "access_key"        // Required
	propertySecretKey       = "secret_key"        // Required
	propertyModel           = "model"             // Optional
	propertyPrompt          = "prompt"            // Optional
	propertyTemperature     = "temperature"       // Optional
	propertyTopP            = "top_p"             // Optional
	propertyMaxTokens       = "max_tokens"        // Optional
	propertyGreeting        = "greeting"          // Optional
	propertyMaxMemoryLength = "max_memory_length" // Optional
)

var (
	memory          []types.Message
	memoryChan      chan types.Message
	maxMemoryLength = 10

	outdateTs atomic.Int64
	wg        sync.WaitGroup
)

func newBedrockLLMExtension(name string) rtego.Extension {
	return &bedrockLLMExtension{}
}

// OnStart will be called when the extension is starting,
// properies can be read here to initialize and start the extension.
// current supported properties:
//   - region (optional)
//   - access_key (required)
//   - secret_key (required)
//   - prompt
//   - temperature
//   - top_p
//   - max_tokens
//   - greeting
func (p *bedrockLLMExtension) OnStart(rte rtego.Rte) {
	slog.Info("OnStart", logTag)

	// prepare configuration
	bedrockLLMConfig := defaultBedrockLLMConfig()

	if accessKey, err := rte.GetPropertyString(propertyAccessKey); err != nil {
		slog.Error(fmt.Sprintf("GetProperty required %s failed, err: %v", propertyAccessKey, err), logTag)
	} else {
		if len(accessKey) > 0 {
			bedrockLLMConfig.AccessKey = accessKey
		}
	}
	if secretKey, err := rte.GetPropertyString(propertySecretKey); err != nil {
		slog.Error(fmt.Sprintf("GetProperty required %s failed, err: %v", propertySecretKey, err), logTag)
	} else {
		if len(secretKey) > 0 {
			bedrockLLMConfig.SecretKey = secretKey
		}
	}

	if model, err := rte.GetPropertyString(propertyModel); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s error:%v", propertyModel, err), logTag)
	} else {
		if len(model) > 0 {
			bedrockLLMConfig.Model = model
		}
	}

	if prompt, err := rte.GetPropertyString(propertyPrompt); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s error:%v", propertyPrompt, err), logTag)
	} else {
		if len(prompt) > 0 {
			bedrockLLMConfig.Prompt = prompt
		}
	}

	if temperature, err := rte.GetPropertyFloat64(propertyTemperature); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyTemperature, err), logTag)
	} else {
		bedrockLLMConfig.Temperature = float32(temperature)
	}

	if topP, err := rte.GetPropertyFloat64(propertyTopP); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyTopP, err), logTag)
	} else {
		bedrockLLMConfig.TopP = float32(topP)
	}

	if maxTokens, err := rte.GetPropertyInt64(propertyMaxTokens); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyMaxTokens, err), logTag)
	} else {
		if maxTokens > 0 {
			bedrockLLMConfig.MaxTokens = int32(maxTokens)
		}
	}

	greeting, err := rte.GetPropertyString(propertyGreeting)
	if err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyGreeting, err), logTag)
	}

	if propMaxMemoryLength, err := rte.GetPropertyInt64(propertyMaxMemoryLength); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyMaxMemoryLength, err), logTag)
	} else {
		if propMaxMemoryLength > 0 {
			maxMemoryLength = int(propMaxMemoryLength)
		}
	}

	// create bedrockLLM instance
	bedrockLLM, err := newBedrockLLM(bedrockLLMConfig)
	if err != nil {
		slog.Error(fmt.Sprintf("newBedrockLLM failed, err: %v", err), logTag)
		return
	}
	slog.Info(fmt.Sprintf("newBedrockLLM succeed with max_tokens: %d, model: %s",
		bedrockLLMConfig.MaxTokens, bedrockLLMConfig.Model), logTag)

	p.bedrockLLM = bedrockLLM

	memoryChan = make(chan types.Message, maxMemoryLength*2)

	// send greeting if available
	if len(greeting) > 0 {
		outputData, _ := rtego.NewData("text_data")
		outputData.SetProperty(dataOutTextDataPropertyText, greeting)
		outputData.SetProperty(dataOutTextDataPropertyTextEndOfSegment, true)
		if err := rte.SendData(outputData); err != nil {
			slog.Error(fmt.Sprintf("greeting [%s] send failed, err: %v", greeting, err), logTag)
		} else {
			slog.Info(fmt.Sprintf("greeting [%s] sent", greeting), logTag)
		}
	}

	rte.OnStartDone()
}

// OnCmd receives cmd from rte graph.
// current supported cmd:
//   - name: flush
//     example:
//     {"name": "flush"}
func (p *bedrockLLMExtension) OnCmd(
	rte rtego.Rte,
	cmd rtego.Cmd,
) {
	cmdName, err := cmd.GetName()
	if err != nil {
		result, fatal := rtego.NewCmdResult(rtego.Error)

		if fatal != nil {
			slog.Error(fmt.Sprintf("OnCmd get name failed, err: %v", err), logTag)
		}
		rte.ReturnResult(result, cmd)
		return
	}
	slog.Info(fmt.Sprintf("OnCmd %s", cmdInFlush), logTag)

	switch cmdName {
	case cmdInFlush:
		outdateTs.Store(time.Now().UnixMicro())

		wg.Wait() // wait for chat completion stream to finish

		// send out
		outCmd, err := rtego.NewCmd(cmdOutFlush)
		if err != nil {
			result, fatal := rtego.NewCmdResult(rtego.Error)

			if fatal != nil {
				slog.Error(fmt.Sprintf("new cmd %s failed, err: %v", cmdOutFlush, err), logTag)
			}
			rte.ReturnResult(result, cmd)
			return
		}
		if err := rte.SendCmd(outCmd, nil); err != nil {
			result, fatal := rtego.NewCmdResult(rtego.Error)

			if fatal != nil {
				slog.Error(fmt.Sprintf("send cmd %s failed, err: %v", cmdOutFlush, err), logTag)
			}
			rte.ReturnResult(result, cmd)
			return
		} else {
			slog.Info(fmt.Sprintf("cmd %s sent", cmdOutFlush), logTag)
		}
	}

	result, _ := rtego.NewCmdResult(rtego.Ok)
	rte.ReturnResult(result, cmd)
}

// OnData receives data from rte graph.
// current supported data:
//   - name: text_data
//     example:
//     {"name": "text_data", "properties": {"text": "hello", "is_final": true}
func (p *bedrockLLMExtension) OnData(
	rte rtego.Rte,
	data rtego.Data,
) {
	// Get isFinal
	isFinal, err := data.GetPropertyBool(dataInTextDataPropertyIsFinal)
	if err != nil {
		slog.Warn(fmt.Sprintf("OnData GetProperty %s failed, err: %v", dataInTextDataPropertyIsFinal, err), logTag)
		return
	}
	if !isFinal { // ignore non-final
		slog.Debug("ignore non-final input", logTag)
		return
	}

	// Get input text
	inputText, err := data.GetPropertyString(dataInTextDataPropertyText)
	if err != nil {
		slog.Error(fmt.Sprintf("OnData GetProperty %s failed, err: %v", dataInTextDataPropertyText, err), logTag)
		return
	}
	if len(inputText) == 0 {
		slog.Debug("ignore empty text", logTag)
		return
	}
	slog.Info(fmt.Sprintf("OnData input text: [%s]", inputText), logTag)

	// prepare memory
	for len(memoryChan) > 0 {
		m, ok := <-memoryChan
		if !ok {
			break
		}
		memory = append(memory, m)
		if len(memory) > maxMemoryLength {
			memory = memory[1:]
		}
	}

	memory = append(memory, types.Message{
		Role: types.ConversationRoleUser,
		Content: []types.ContentBlock{
			&types.ContentBlockMemberText{
				Value: inputText,
			},
		},
	})

	if len(memory) > maxMemoryLength {
		memory = memory[1:]
	}

	// start goroutine to request and read responses from bedrock
	wg.Add(1)
	go func(startTime time.Time, inputText string, memory []types.Message) {
		defer wg.Done()
		slog.Info(fmt.Sprintf("getConvserseStream for input text: [%s] memory: %v", inputText, memory), logTag)

		// Get result from ai
		resp, err := p.bedrockLLM.getConverseStream(memory)
		if err != nil {
			slog.Error(fmt.Sprintf("getConvserseStream for input text: [%s] failed, err: %v", inputText, err), logTag)
			return
		}
		// defer func() {
		// 	if resp != nil { // Close stream object
		// 		resp.Close()
		// 	}
		// }()
		slog.Debug(fmt.Sprintf("getConvserseStream start to recv for input text: [%s]", inputText), logTag)

		var sentence, fullContent string
		var firstSentenceSent bool
		for event := range resp.GetStream().Events() {
			if startTime.UnixMicro() < outdateTs.Load() { // Check whether to interrupt
				slog.Info(fmt.Sprintf("GetChatCompletionsStream recv interrupt and flushing for input text: [%s], startTs: %d, outdateTs: %d",
					inputText, startTime.UnixMicro(), outdateTs.Load()), logTag)
				break
			}
			var content string

			switch v := event.(type) {
			// case *types.ConverseStreamOutputMemberMessageStart:
			// msg.Role = v.Value.Role
			case *types.ConverseStreamOutputMemberContentBlockDelta:
				textResponse := v.Value.Delta.(*types.ContentBlockDeltaMemberText)
				content = textResponse.Value

			case *types.UnknownUnionMember:
				fmt.Println("unknown tag:", v.Tag)
			}

			// chatCompletions, err := resp.Recv()
			// if errors.Is(err, io.EOF) {
			// 	slog.Debug(fmt.Sprintf("GetChatCompletionsStream recv for input text: [%s], io.EOF break", inputText), logTag)
			// 	break
			// }

			// if len(chatCompletions.Choices) > 0 && chatCompletions.Choices[0].Delta.Content != "" {
			// 	content = chatCompletions.Choices[0].Delta.Content
			// }
			fullContent += content

			for {
				// feed content and check whether sentence is available
				var sentenceIsFinal bool
				sentence, content, sentenceIsFinal = parseSentence(sentence, content)
				if len(sentence) == 0 || !sentenceIsFinal {
					slog.Debug(fmt.Sprintf("sentence %s is empty or not final", sentence), logTag)
					break
				}
				slog.Debug(fmt.Sprintf("GetChatCompletionsStream recv for input text: [%s] got sentence: [%s]", inputText, sentence), logTag)

				// send sentence
				outputData, err := rtego.NewData("text_data")
				if err != nil {
					slog.Error(fmt.Sprintf("NewData failed, err: %v", err), logTag)
					break
				}
				outputData.SetProperty(dataOutTextDataPropertyText, sentence)
				outputData.SetProperty(dataOutTextDataPropertyTextEndOfSegment, false)
				if err := rte.SendData(outputData); err != nil {
					slog.Error(fmt.Sprintf("GetChatCompletionsStream recv for input text: [%s] send sentence [%s] failed, err: %v", inputText, sentence, err), logTag)
					break
				} else {
					slog.Info(fmt.Sprintf("GetChatCompletionsStream recv for input text: [%s] sent sentence [%s]", inputText, sentence), logTag)
				}
				sentence = ""

				if !firstSentenceSent {
					firstSentenceSent = true
					slog.Info(fmt.Sprintf("GetChatCompletionsStream recv for input text: [%s] first sentence sent, first_sentency_latency %dms",
						inputText, time.Since(startTime).Milliseconds()), logTag)
				}
			}
		}

		// remember response as assistant content in memory
		memoryChan <- types.Message{
			Role: types.ConversationRoleAssistant,
			Content: []types.ContentBlock{
				&types.ContentBlockMemberText{
					Value: fullContent,
				},
			},
		}

		// send end of segment
		outputData, _ := rtego.NewData("text_data")
		outputData.SetProperty(dataOutTextDataPropertyText, sentence)
		outputData.SetProperty(dataOutTextDataPropertyTextEndOfSegment, true)
		if err := rte.SendData(outputData); err != nil {
			slog.Error(fmt.Sprintf("GetChatCompletionsStream for input text: [%s] end of segment with sentence [%s] send failed, err: %v", inputText, sentence, err), logTag)
		} else {
			slog.Info(fmt.Sprintf("GetChatCompletionsStream for input text: [%s] end of segment with sentence [%s] sent", inputText, sentence), logTag)
		}
	}(time.Now(), inputText, append([]types.Message{}, memory...))
}

func init() {
	slog.Info("init")

	// Register addon
	rtego.RegisterAddonAsExtension(
		"bedrock_llm",
		rtego.NewDefaultExtensionAddon(newBedrockLLMExtension),
	)
}

File: agents/addon/extension/bedrock_llm_python/bedrock_llm_extension.py
from .bedrock_llm import BedrockLLM, BedrockLLMConfig
from datetime import datetime
from threading import Thread
from rte_runtime_python import (
    Addon,
    Extension,
    register_addon_as_extension,
    Rte,
    Cmd,
    Data,
    StatusCode,
    CmdResult,
    MetadataInfo,
    RTE_PIXEL_FMT,
)
from rte_runtime_python.image_frame import ImageFrame
from PIL import Image, ImageFilter
from .log import logger


CMD_IN_FLUSH = "flush"
CMD_OUT_FLUSH = "flush"
DATA_IN_TEXT_DATA_PROPERTY_TEXT = "text"
DATA_IN_TEXT_DATA_PROPERTY_IS_FINAL = "is_final"
DATA_OUT_TEXT_DATA_PROPERTY_TEXT = "text"
DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT = "end_of_segment"

PROPERTY_REGION = "region"  # Optional
PROPERTY_ACCESS_KEY = "access_key"  # Optional
PROPERTY_SECRET_KEY = "secret_key"  # Optional
PROPERTY_MODEL = "model"  # Optional
PROPERTY_PROMPT = "prompt"  # Optional
PROPERTY_TEMPERATURE = "temperature"  # Optional
PROPERTY_TOP_P = "top_p"  # Optional
PROPERTY_MAX_TOKENS = "max_tokens"  # Optional
PROPERTY_GREETING = "greeting"  # Optional
PROPERTY_MAX_MEMORY_LENGTH = "max_memory_length"  # Optional

def get_current_time():
    # Get the current time
    start_time = datetime.now()
    # Get the number of microseconds since the Unix epoch
    unix_microseconds = int(start_time.timestamp() * 1_000_000)
    return unix_microseconds

def is_punctuation(char):
    if char in [',', '，', '.', '。', '?', '？', '!', '！']:
        return True
    return False

def parse_sentence(sentence, content):
    remain = ""
    found_punc = False

    for char in content:
        if not found_punc:
            sentence += char
        else:
            remain += char

        if not found_punc and is_punctuation(char):
            found_punc = True

    return sentence, remain, found_punc

class BedrockLLMExtension(Extension):
    memory = []
    max_memory_length = 10
    outdate_ts = 0
    bedrock_llm = None

    def on_init(self, rte: Rte, manifest: MetadataInfo, property: MetadataInfo) -> None:
        logger.info("BedrockLLMExtension on_init")
        rte.on_init_done(manifest, property)

    def on_start(self, rte: Rte) -> None:
        logger.info("BedrockLLMExtension on_start")
        # Prepare configuration
        bedrock_llm_config = BedrockLLMConfig.default_config()

        for optional_str_param in [
            PROPERTY_REGION, PROPERTY_ACCESS_KEY, PROPERTY_SECRET_KEY,
            PROPERTY_MODEL, PROPERTY_PROMPT]:
            try:
                value = rte.get_property_string(optional_str_param).strip()
                if value:
                    bedrock_llm_config.__setattr__(optional_str_param, value)
            except Exception as err:
                logger.debug(f"GetProperty optional {optional_str_param} failed, err: {err}. Using default value: {bedrock_llm_config.__getattribute__(optional_str_param)}")

        for optional_float_param in [PROPERTY_TEMPERATURE, PROPERTY_TOP_P]:
            try:
                value = rte.get_property_float(optional_float_param)
                if value:
                    bedrock_llm_config.__setattr__(optional_float_param, value)
            except Exception as err:
                logger.debug(f"GetProperty optional {optional_float_param} failed, err: {err}. Using default value: {bedrock_llm_config.__getattribute__(optional_float_param)}")

        try:
            max_tokens = rte.get_property_int(PROPERTY_MAX_TOKENS)
            if max_tokens > 0:
                bedrock_llm_config.max_tokens = int(max_tokens)
        except Exception as err:
            logger.debug(f"GetProperty optional {PROPERTY_MAX_TOKENS} failed, err: {err}. Using default value: {bedrock_llm_config.max_tokens}")

        try:
            greeting = rte.get_property_string(PROPERTY_GREETING)
        except Exception as err:
            logger.debug(f"GetProperty optional {PROPERTY_GREETING} failed, err: {err}.")

        try:
            prop_max_memory_length = rte.get_property_int(PROPERTY_MAX_MEMORY_LENGTH)
            if prop_max_memory_length > 0:
                self.max_memory_length = int(prop_max_memory_length)
        except Exception as err:
            logger.debug(f"GetProperty optional {PROPERTY_MAX_MEMORY_LENGTH} failed, err: {err}.")

        # Create bedrockLLM instance
        try:
            self.bedrock_llm = BedrockLLM(bedrock_llm_config)
            logger.info(f"newBedrockLLM succeed with max_tokens: {bedrock_llm_config.max_tokens}, model: {bedrock_llm_config.model}")
        except Exception as err:
            logger.exception(f"newBedrockLLM failed, err: {err}")

        # Send greeting if available
        if greeting:
            try:
                output_data = Data.create("text_data")
                output_data.set_property_string(DATA_OUT_TEXT_DATA_PROPERTY_TEXT, greeting)
                output_data.set_property_bool(DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT, True)
                rte.send_data(output_data)
                logger.info(f"greeting [{greeting}] sent")
            except Exception as err:
                logger.info(f"greeting [{greeting}] send failed, err: {err}")
        rte.on_start_done()

    def on_stop(self, rte: Rte) -> None:
        logger.info("BedrockLLMExtension on_stop")
        rte.on_stop_done()

    def on_deinit(self, rte: Rte) -> None:
        logger.info("BedrockLLMExtension on_deinit")
        rte.on_deinit_done()

    def on_cmd(self, rte: Rte, cmd: Cmd) -> None:
        logger.info("BedrockLLMExtension on_cmd")
        cmd_json = cmd.to_json()
        logger.info("BedrockLLMExtension on_cmd json: " + cmd_json)

        cmd_name = cmd.get_name()

        if cmd_name == CMD_IN_FLUSH:
            self.outdate_ts = get_current_time()
            cmd_out = Cmd.create(CMD_OUT_FLUSH)
            rte.send_cmd(cmd_out, None)
            logger.info(f"BedrockLLMExtension on_cmd sent flush")
        else:
            logger.info(f"BedrockLLMExtension on_cmd unknown cmd: {cmd_name}")
            cmd_result = CmdResult.create(StatusCode.ERROR)
            cmd_result.set_property_string("detail", "unknown cmd")
            rte.return_result(cmd_result, cmd)
            return

        cmd_result = CmdResult.create(StatusCode.OK)
        cmd_result.set_property_string("detail", "success")
        rte.return_result(cmd_result, cmd)

    def on_image_frame(self, rte: Rte, image_frame: ImageFrame) -> None:
        logger.info("BedrockLLMExtension on_cmd")

    def on_data(self, rte: Rte, data: Data) -> None:
        """
        on_data receives data from rte graph.
        current supported data:
          - name: text_data
            example:
            {name: text_data, properties: {text: "hello"}
        """
        logger.info(f"BedrockLLMExtension on_data")

        # Assume 'data' is an object from which we can get properties
        try:
            is_final = data.get_property_bool(DATA_IN_TEXT_DATA_PROPERTY_IS_FINAL)
            if not is_final:
                logger.info("ignore non-final input")
                return
        except Exception as err:
            logger.info(f"OnData GetProperty {DATA_IN_TEXT_DATA_PROPERTY_IS_FINAL} failed, err: {err}")
            return

        # Get input text
        try:
            input_text = data.get_property_string(DATA_IN_TEXT_DATA_PROPERTY_TEXT)
            if not input_text:
                logger.info("ignore empty text")
                return
            logger.info(f"OnData input text: [{input_text}]")
        except Exception as err:
            logger.info(f"OnData GetProperty {DATA_IN_TEXT_DATA_PROPERTY_TEXT} failed, err: {err}")
            return

        # Prepare memory. A conversation must alternate between user and assistant roles
        while len(self.memory):
            if len(self.memory) > self.max_memory_length:
                logger.debug(f'pop out first message, reason: memory length limit: `{self.memory[0]}`')
                self.memory.pop(0)
            elif self.memory[0]['role'] == 'assistant':
                logger.debug(f'pop out first message, reason: messages can not start with assistant: `{self.memory[0]}`')
                self.memory.pop(0)
            else:
                break

        if len(self.memory) and self.memory[-1]['role'] == 'user':
            # if last user input got empty response, append current user input.
            logger.debug(f'found last message with role `user`, will append this input into last user input')
            self.memory[-1]['content'].append(
                {'text': input_text}
            )
        else:
            self.memory.append({"role": "user", "content": [
                {'text': input_text}
            ]})

        def converse_stream_worker(start_time, input_text, memory):
            try:
                logger.info(f"GetConverseStream for input text: [{input_text}] memory: {memory}")

                # Get result from Bedrock
                resp = self.bedrock_llm.get_converse_stream(memory)
                if resp is None or resp.get('stream') is None:
                    logger.info(f"GetConverseStream for input text: [{input_text}] failed")
                    return

                stream = resp.get('stream')
                sentence = ""
                full_content = ""
                first_sentence_sent = False

                for event in stream:
                    if start_time < self.outdate_ts:
                        logger.info(f"GetConverseStream recv interrupt and flushing for input text: [{input_text}], startTs: {start_time}, outdateTs: {self.outdate_ts}")
                        break

                    if 'contentBlockDelta' in event:
                        delta_types = event['contentBlockDelta']['delta'].keys()
                        # ignore other types of content: e.g toolUse
                        if 'text' in delta_types:
                            content = event['contentBlockDelta']['delta']['text']
                    elif 'internalServerException' in event or 'modelStreamErrorException' in event \
                        or 'throttlingException' in event or 'validationException' in event:
                        logger.error(f"GetConverseStream Error occured: {event}")
                        break
                    else:
                        # ingore other events
                        continue

                    full_content += content

                    while True:
                        sentence, content, sentence_is_final = parse_sentence(sentence, content)
                        if len(sentence) == 0 or not sentence_is_final:
                            logger.info(f"sentence {sentence} is empty or not final")
                            break
                        logger.info(f"GetConverseStream recv for input text: [{input_text}] got sentence: [{sentence}]")

                        # send sentence
                        try:
                            output_data = Data.create("text_data")
                            output_data.set_property_string(DATA_OUT_TEXT_DATA_PROPERTY_TEXT, sentence)
                            output_data.set_property_bool(DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT, False)
                            rte.send_data(output_data)
                            logger.info(f"GetConverseStream recv for input text: [{input_text}] sent sentence [{sentence}]")
                        except Exception as err:
                            logger.info(f"GetConverseStream recv for input text: [{input_text}] send sentence [{sentence}] failed, err: {err}")
                            break

                        sentence = ""
                        if not first_sentence_sent:
                            first_sentence_sent = True
                            logger.info(f"GetConverseStream recv for input text: [{input_text}] first sentence sent, first_sentence_latency {get_current_time() - start_time}ms")

                if len(full_content.strip()):
                    # remember response as assistant content in memory
                    memory.append({"role": "assistant", "content": [{"text": full_content}]})
                else:
                    # can not put empty model response into memory
                    logger.error(f"GetConverseStream recv for input text: [{input_text}] failed: empty response [{full_content}]")
                    return

                # send end of segment
                try:
                    output_data = Data.create("text_data")
                    output_data.set_property_string(DATA_OUT_TEXT_DATA_PROPERTY_TEXT, sentence)
                    output_data.set_property_bool(DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT, True)
                    rte.send_data(output_data)
                    logger.info(f"GetConverseStream for input text: [{input_text}] end of segment with sentence [{sentence}] sent")
                except Exception as err:
                    logger.info(f"GetConverseStream for input text: [{input_text}] end of segment with sentence [{sentence}] send failed, err: {err}")

            except Exception as e:
                logger.info(f"GetConverseStream for input text: [{input_text}] failed, err: {e}")

        # Start thread to request and read responses from OpenAI
        start_time = get_current_time()
        thread = Thread(target=converse_stream_worker, args=(start_time, input_text, self.memory))
        thread.start()
        logger.info(f"BedrockLLMExtension on_data end")


@register_addon_as_extension("bedrock_llm_python")
class BedrockLLMExtensionAddon(Addon):
    def on_init(self, rte: Rte, manifest, property) -> None:
        logger.info("BedrockLLMExtensionAddon on_init")
        rte.on_init_done(manifest, property)
        return

    def on_create_instance(self, rte: Rte, addon_name: str, context) -> None:
        logger.info("on_create_instance")
        rte.on_create_instance_done(BedrockLLMExtension(addon_name), context)

    def on_deinit(self, rte: Rte) -> None:
        logger.info("BedrockLLMExtensionAddon on_deinit")
        rte.on_deinit_done()
        return
